File: 1. Two Sum/Java/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class Main{
    // Taking input using input.txt file
    public static void main(String[] args) throws Exception {
        // Reading Input File
        BufferedReader br = new BufferedReader(new FileReader("../input.txt"));
        Solution sol = new Solution();

        int passed = 0;
        int total = 0;
        
        String line;
        while ((line = br.readLine()) != null) {
            line = line.trim();
            if (line.isEmpty()){
                continue;
            }

            if (line.length() < 2){
                break;
            }

            // Reading Array
            line = line.substring(1, line.length() - 1);
            String[] parts = line.split(",");

            int[] nums = new int[parts.length];
            for (int i = 0; i < parts.length; i++){
                nums[i] = Integer.parseInt(parts[i].trim());
            }

            // Reading target
            int target = Integer.parseInt(br.readLine().trim());

            // Reading expected output
            String expectedOutput = br.readLine().trim();
            expectedOutput = expectedOutput.substring(1, expectedOutput.length() - 1);
            String[] expectedOut = expectedOutput.split(",");
            int[] expected = new int[expectedOut.length];
            for (int i = 0; i < expectedOut.length; i++){
                expected[i] = Integer.parseInt(expectedOut[i].trim());
            }

            total++;
            int[] result = sol.twoSum(nums, target);
            
            if (Arrays.equals(result, expected)){
                passed++;
            }
            else{
                System.out.println("Test Case " + total + ": FAILED");
                System.out.println("Input: num -> " + Arrays.toString(nums) + ", target -> " + target);
                System.out.println("Expected: " + Arrays.toString(expected));
                System.out.println("Your Output: " + Arrays.toString(result));
                br.close();
                System.exit(0);
            }
        }

        br.close();
        System.out.println("Result: " + passed + " / " + total + " test cases passed");
    }
}

class Solution{
    /*
    *
    * Approach
    * We will be using a hashmap to store a value and its index as key value pairs
    * we target - current number is found in the map,
    * we have found our answer
    * otherwise simply put the current number along with its index in the
    * 
    * Time Complexity - O(n), Space Complexity - O(n)
    * 
    */
    public int[] twoSum(int[] nums, int target){
        Map<Integer, Integer> map = new HashMap<>(); // Creating an empty HashMap
        for (int i = 0; i < nums.length; i++){
            int idx = map.getOrDefault(target - nums[i], -1); // Checking whether the needed result exists in our HashMap
            if (idx != -1){
                return new int[]{idx, i}; // Returning the index
            }
            map.put(nums[i], 0);
        }
        return new int[]{};
    }
}

File: 1. Two Sum/C++/main.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{
    public:
        vector<int> twoSum(vector<int>& nums, int target){
            unordered_map<int, int> map;
            for (int i = 0; i < nums.size(); i++){
                int val = target - nums[i];
                if (map.find(val) != map.end()){
                    return {map[val], i};
                }
                map.insert({nums[i], i});
            }
            return {};
        }
};

int main(){
    ifstream fin("../input.txt");
    if (!fin.is_open()){
        cerr << "Failed to open input.txt" << endl;
        return 1;
    }
    
    Solution sol;

    int passed = 0;
    int total = 0;

    string line;
    while (getline(fin, line)){
        line.erase(remove_if(line.begin(), line.end(), ::isspace), line.end());

        if (line.empty()){
            continue;
        }

        if (line.size() < 2){
            break;
        }

        // Reading Array
        line = line.substr(1, line.size() - 2);
        vector<int> nums;
        if (!line.empty()){
            stringstream ss(line);
            string token;
            while (getline(ss, token, ',')){
                nums.push_back(stoi(token));
            }
        }

        // Reading target
        string targetLine;
        if (!getline(fin, targetLine)){
            break;
        }
        int target = stoi(targetLine);

        // Reading expected output
        string expectedLine;
        if (!getline(fin, expectedLine)){
            break;
        }
        expectedLine.erase(remove_if(expectedLine.begin(), expectedLine.end(), ::isspace), expectedLine.end());
        expectedLine = expectedLine.substr(1, expectedLine.size() - 2);

        vector<int> expected;
        if (!expectedLine.empty()){
            stringstream ss(expectedLine);
            string token;
            while (getline(ss, token, ',')){
                expected.push_back(stoi(token));
            }
        }

        total++;
        vector<int> result = sol.twoSum(nums, target);
        if (result == expected){
            passed++;
        }
        else{
            cout << "Test Case " << total << ": FAILED" << endl;
            cout << "Input: nums = [";
            for (int i = 0; i < nums.size(); i++){
                cout << nums[i] << (i + 1 < nums.size() ? "," : "");
            }
            cout << "], target = " << target << endl;

            cout << "Expected: [";
            for (int i = 0; i < expected.size(); i++){
                cout << expected[i] << (i + 1 < expected.size() ? "," : "");
            }
            cout << "]" << endl;

            cout << "Your Output: [";
            for (int i = 0; i < result.size(); i++){
                cout << result[i] << (i + 1 < result.size() ? "," : "");
            }
            cout << "]" << endl;
            return 1;
        }
    }

    cout << "Result: " << passed << " / " << total << " test cases passed" << endl;
    return 0;
}